File: frontend/src/lib/permissions.ts
// Capability-based permissions, mirroring app/core/permissions.py on the backend.
// Add a new role? Define its capability set here — that's the only file that needs to change.
// Components should check capabilities (useCapability / hasCapability), not role names.

import { useAuthStore } from '@/store/auth'

export type Capability =
  | 'manage_users'
  | 'manage_projects'
  | 'manage_project_members'
  | 'manage_project_resources'
  | 'manage_schemas'
  | 'upload_extraction_jobs'
  | 'submit_work'
  | 'review_submissions'
  | 'review_records'
  | 'view_all_projects'

const ALL_CAPABILITIES: Capability[] = [
  'manage_users', 'manage_projects', 'manage_project_members', 'manage_project_resources',
  'manage_schemas', 'upload_extraction_jobs', 'submit_work', 'review_submissions',
  'review_records', 'view_all_projects',
]

export const ROLE_CAPABILITIES: Record<string, Capability[]> = {
  org_admin: ALL_CAPABILITIES,
  project_admin: [
    'manage_project_members', 'manage_project_resources', 'manage_schemas',
    'upload_extraction_jobs', 'review_submissions', 'submit_work',
  ],
  qa_lead: ['review_submissions', 'review_records'],
  pipeline_operator: ['upload_extraction_jobs', 'submit_work'],
  reviewer: ['review_submissions', 'review_records'],
  read_only: [],
}

export function capabilitiesForRoles(roles: string[]): Set<Capability> {
  const caps = new Set<Capability>()
  for (const role of roles) {
    for (const cap of ROLE_CAPABILITIES[role] ?? []) caps.add(cap)
  }
  return caps
}

export function hasCapability(roles: string[], capability: Capability): boolean {
  return capabilitiesForRoles(roles).has(capability)
}

/** Use inside components: const canManageUsers = useCapability('manage_users') */
export function useCapability(capability: Capability): boolean {
  const user = useAuthStore(s => s.user)
  if (!user) return false
  return hasCapability(user.roles, capability)
}

// ─── Project-scoped checks ──────────────────────────────────────────────────
// A user's role on a specific project (via project membership) is a separate
// dimension from their global role — someone can be globally "reviewer" but
// locally "project_admin" on one project. These helpers take the project's
// member list (already returned alongside project detail, see ProjectMember
// type) plus the current user, for screens that need project-scoped checks.

export interface ProjectMemberLite {
  user_id: string
  role: string
}

export function isProjectAdmin(globalRoles: string[], userId: string, members: ProjectMemberLite[]): boolean {
  if (globalRoles.includes('org_admin')) return true
  return members.some(m => m.user_id === userId && m.role === 'project_admin')
}

export function canReviewProject(globalRoles: string[], userId: string, members: ProjectMemberLite[]): boolean {
  if (globalRoles.includes('org_admin') || globalRoles.includes('qa_lead')) return true
  return members.some(m => m.user_id === userId && (m.role === 'project_admin' || m.role === 'reviewer'))
}

// ─── Role display metadata ────────────────────────────────────────────────────
// Use these everywhere in the UI instead of hardcoded strings, so renaming a
// role in future only requires changing it here.

export const ROLE_META: Record<string, { label: string; description: string; color: string }> = {
  org_admin: {
    label: 'Admin',
    description: 'Full access — manages users, projects, schemas, and can download exports.',
    color: 'red',
  },
  project_admin: {
    label: 'Project Admin',
    description: 'Manages one project — assigns extractors and reviewers, sees project stats.',
    color: 'amber',
  },
  qa_lead: {
    label: 'QA Lead',
    description: 'Cross-project review access — can approve or reject records in any project.',
    color: 'blue',
  },
  pipeline_operator: {
    label: 'Extractor',
    description: 'Uploads source files and fixes validation errors on assigned sources.',
    color: 'green',
  },
  reviewer: {
    label: 'Reviewer',
    description: 'Reviews extracted records against the source website on assigned sources.',
    color: 'purple',
  },
  read_only: {
    label: 'Read Only',
    description: 'Can view projects and sources but cannot make any changes.',
    color: 'gray',
  },
}

export function getRoleLabel(role: string): string {
  return ROLE_META[role]?.label ?? role
}

File: backend/app/core/permissions.py
"""
Capability-based permissions layer.

Role checks scattered as `require_roles("org_admin", "project_admin")` get
brittle as new roles are added, since every new role means re-checking every
call site that cares about that capability. This module defines what each
GLOBAL role can do once; routes check capabilities, not role names.

Note: this only models global roles (UserRoleAssignment / user.roles). Project-
scoped checks ("is this user a reviewer on THIS specific project") still need
the per-project logic in project_resources.py / work_submissions.py, since a
user's project-level role (ProjectMember.role) is a separate dimension from
their global role — a user can be globally "reviewer" but locally
"project_admin" on one specific project, or vice versa. Use this module for
global, not-project-scoped checks (e.g. "can manage users platform-wide").

New routes use this where the check is global-only. Existing routes
(projects, users, jobs, schemas) still use the original `require_roles`
dependency in app/core/security.py — migrate them here incrementally, since
swapping every working endpoint in one pass is needless risk.
"""
from enum import Enum
from fastapi import Depends, HTTPException, status
from app.core.security import get_current_user


class Capability(str, Enum):
    MANAGE_USERS = "manage_users"
    MANAGE_PROJECTS = "manage_projects"
    MANAGE_PROJECT_MEMBERS = "manage_project_members"
    MANAGE_PROJECT_RESOURCES = "manage_project_resources"
    MANAGE_SCHEMAS = "manage_schemas"
    UPLOAD_EXTRACTION_JOBS = "upload_extraction_jobs"
    SUBMIT_WORK = "submit_work"
    REVIEW_SUBMISSIONS = "review_submissions"
    REVIEW_RECORDS = "review_records"
    VIEW_ALL_PROJECTS = "view_all_projects"


# Add a new role to this platform? Define its capability set here — nowhere else.
ROLE_CAPABILITIES: dict[str, set] = {
    "org_admin": set(Capability),  # everything
    "project_admin": {
        Capability.MANAGE_PROJECT_MEMBERS,
        Capability.MANAGE_PROJECT_RESOURCES,
        Capability.MANAGE_SCHEMAS,
        Capability.UPLOAD_EXTRACTION_JOBS,
        Capability.REVIEW_SUBMISSIONS,
        Capability.REVIEW_RECORDS,
        Capability.SUBMIT_WORK,
        Capability.VIEW_ALL_PROJECTS,
    },
    "qa_lead": {
        Capability.REVIEW_SUBMISSIONS,
        Capability.REVIEW_RECORDS,
        Capability.SUBMIT_WORK,
        Capability.UPLOAD_EXTRACTION_JOBS,
    },
    "pipeline_operator": {
        Capability.UPLOAD_EXTRACTION_JOBS,
        Capability.SUBMIT_WORK,
    },
    "reviewer": {
        Capability.REVIEW_SUBMISSIONS,
        Capability.REVIEW_RECORDS,
    },
    "read_only": set(),
}


def user_capabilities(user) -> set:
    caps: set = set()
    for r in user.roles:
        caps |= ROLE_CAPABILITIES.get(r.role.value, set())
    return caps


def has_capability(user, capability: Capability) -> bool:
    return capability in user_capabilities(user)


def require_capability(*capabilities: Capability):
    """Dependency factory — require at least one of the given capabilities, globally."""
    async def _checker(current_user=Depends(get_current_user)):
        if not user_capabilities(current_user).intersection(set(capabilities)):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Insufficient permissions. Requires one of: {[c.value for c in capabilities]}",
            )
        return current_user
    return _checker